### PR TITLE
[Performance] use sort-based top-k/top-p in non-reduce mode for consistent performance

### PR DESCRIPTION
**Pull request by JC-ut0**
## Problem

In vllm-ascend, the top-k/top-p sampling operator on A2/A3 NPU uses `torch_npu.npu_top_k_top_p` (a fused C++ operator) in non-reduce_sample mode. Benchmark testing revealed that this operator **degrades severely** when:

- `k` is large (e.g., k=128000, k=vocab_size=152000)
- The batch contains **mixed k values** (different requests have different top_k)

In realistic production traffic where each request has different sampling parameters, `npu_top_k_top_p` takes **19-27ms** per sampling step, while an alternative sort-based implementation using standard PyTorch ops takes only **~1ms** — a **17-25x performance gap**.

## Root Cause

`npu_top_k_top_p` uses a partial selection algorithm whose complexity scales with `k`. When `k` is small (e.g., k=1), it is fast (~0.5ms). But when `k` is large or the batch contains requests with large k (including "unset" requests where upstream vLLM fills `k=vocab_size`), the operator's internal algorithm becomes much slower than a simple full sort + mask approach.

In contrast, `_apply_top_k_top_p_pytorch` uses `softmax -> sort -> gather -> mask`, which is O(V log V) regardless of k. The `torch.sort` kernel on NPU is highly optimized, and the subsequent operations pipeline well across multiple kernel launches with no host-device synchronization.

## Fix

In `vllm_ascend/sample/sampler.py`, function `_apply_top_k_top_p_torch_npu`:

**Before (non-reduce mode):**
```python
if p is None and k is None:
    return logits
return torch_npu.npu_top_k_top_p(logits, k=k, p=p)
```

**After:**
```python
# Non-reduce_sample mode: use sort-based pytorch implementation.
# npu_top_k_top_p degrades severely (5-28ms) when k is large or batch
# contains mixed k values, while sort+mask is consistently ~1ms.
return _apply_top_k_top_p_pytorch(logits, k, p)
```

The reduce_sample mode is **unchanged** — it still uses `npu_top_k_top_p` on all_gather results, because that path requires the fused operator to process gathered candidate tensors.

### Environment
- Hardware: Ascend A3
- batch=32, vocab=152000, dtype=float32
- warmup=5, iters=20
- Times are median (ms)

### Uniform Cases (all requests share same k/p)

| Case | A:pt_npu | B:npu | ratio | best |
|---|---|---|---|---|
| k=1,p=None | 0.654 | 0.519 | 1.26x | B:npu |
| k=1,p=1 | 1.099 | 0.538 | 2.04x | B:npu |
| k=1,p=0.95 | 1.101 | 0.536 | 2.06x | B:npu |
| k=1,p=0.5 | 1.110 | 0.544 | 2.04x | B:npu |
| k=None,p=None | 0.014 | 0.021 | 0.67x | A:pt_npu |
| k=None,p=1 | 1.044 | 27.699 | 0.04x | A:pt_npu |
| k=None,p=0.95 | 1.029 | 20.719 | 0.05x | A:pt_npu |
| k=None,p=0.5 | 1.024 | 5.073 | 0.20x | A:pt_npu |
| k=128000,p=None | 0.643 | 22.195 | 0.03x | A:pt_npu |
| k=128000,p=1 | 1.117 | 22.263 | 0.05x | A:pt_npu |
| k=128000,p=0.95 | 1.121 | 18.431 | 0.06x | A:pt_npu |
| k=128000,p=0.5 | 1.105 | 6.104 | 0.18x | A:pt_npu |
| k=4096,p=None | 0.647 | 1.205 | 0.54x | A:pt_npu |
| k=4096,p=1 | 1.104 | 1.271 | 0.87x | A:pt_npu |
| k=4096,p=0.95 | 1.109 | 1.194 | 0.93x | A:pt_npu |
| k=4096,p=0.5 | 1.098 | 0.821 | 1.34x | B:npu |

- A = `_apply_top_k_top_p_pytorch` (sort+mask, the new default)
- B = `npu_top_k_top_p` (fused NPU operator, the old default)
- ratio = A/B (>1 means B is faster)

Key observations:
- B only wins when k=1 (fast partial selection for tiny k)
- B is catastrophically slow when k >= 128000 or k=None with p set (5-28ms)
- A is consistently 0.6-1.1ms regardless of k

### Mixed Batch (per-request k/p all different, realistic production)

| Pattern | A:pt_npu | B:npu | ratio | best |
|---|---|---|---|---|
| typical_prod | 1.097 | 27.379 | 0.04x | A:pt_npu |
| random_mix | 1.104 | 26.777 | 0.04x | A:pt_npu |
| half_k1_half_unrestricted | 1.127 | 19.708 | 0.06x | A:pt_npu |
| mostly_small_k_one_unrestricted | 1.117 | 19.188 | 0.06x | A:pt_npu |

- A wins all mixed batch patterns by **17-25x**
- B is slow because even one request with large k forces the operator to do heavy work for that row

### Correctness

```
pytorch(NPU) vs npu_direct(NPU) valid-position match: True
```

Both implementations produce identical sets of valid (non -inf) positions.

## Why Not Hybrid (choose by k value)?

Considered but rejected:
1. Checking `k.max().item()` to decide which path to use introduces a host-device sync (~0.5-2ms per call), which could negate the benefit
2. In mixed batch, `k.max()` is always inflated by the request with k=vocab_size, so the hybrid would always fall back to pytorch anyway — wasting a sync for nothing
3. The only scenario where B is faster (k=1) is negligible: 0.5ms vs 0.65ms, a 0.15ms difference. In production, k=1 (greedy) typically means top-k is not set (k=None), which skips the sampler entirely

## Changes

| File | Change |
|---|---|
| `vllm_ascend/sample/sampler.py` | Non-reduce mode delegates to `_apply_top_k_top_p_pytorch` instead of `npu_top_k_top_p` |
| `tests/ut/sample/test_sampler.py` | Rewritten to match current implementation. Tests non-reduce mode correctness (top-k, top-p, mixed k, k=V, p=1.0) and reduce mode behavior (mocked) |
| `benchmarks/ops/bench_topk_topp_sampler.py` | New A/B benchmark comparing pytorch vs npu_top_k_top_p with 16 uniform cases + 4 mixed batch patterns |

## How to Reproduce the Benchmark

On an NPU machine:

```bash
python benchmarks/ops/bench_topk_topp_sampler.py
# or without visualization:
python benchmarks/ops/bench_topk_topp_sampler.py --no-plot
```

- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

File: vllm_ascend/sample/sampler.py [hunk 1]
```diff
@@ -259,9 +259,10 @@ def _apply_top_k_top_p_torch_npu(
             gathered_vals = torch_npu.npu_top_k_top_p(gathered_vals, k=k, p=p)
         return gathered_vals, gathered_idx
 
-    if p is None and k is None:
-        return logits
-    return torch_npu.npu_top_k_top_p(logits, k=k, p=p)
+    # Non-reduce_sample mode: use sort-based pytorch implementation.
+    # npu_top_k_top_p degrades severely (5-28ms) when k is large or batch
+    # contains mixed k values, while sort+mask is consistently ~1ms.
+    return _apply_top_k_top_p_pytorch(logits, k, p)
```
[the remark this answers is not in the document]
Comment (JC-ut0) [reply vllm_ascend/sample/sampler.py:267]: Thank you for the analysis. However, k > vocab_size cannot occur in practice — upstream vLLM converts k to vocab_size before reaching this function, so the negative index scenario never happens. I've reverted the clamp (c7f7675) to keep the code clean.

File: tests/ut/sample/test_sampler.py [hunk 2]
```diff
@@ -8,3 +17,169 @@ def test_init_with_raw_logprobs(self):
...
+    def test_top_k_eq_vocab_no_filtering(self, mock_config):
+        """When k == V (no restriction), all logits should be kept."""
+        mock_config.return_value.enable_reduce_sample = False
+        logits = torch.tensor([[1.0, 3.0, 2.0]])
+        k = torch.tensor([3])  # k == V
+        result = _apply_top_k_top_p_torch_npu(logits, k, None)
+        # No filtering: all values kept (no -inf)
+        assert not torch.isinf(result).any()
+
+    @patch("vllm_ascend.sample.sampler.get_ascend_config")
```
[the remark this answers is not in the document]
Comment (JC-ut0) [reply tests/ut/sample/test_sampler.py]: Since k > vocab_size does not occur in practice (upstream vLLM converts it to vocab_size beforehand), I've reverted the test as well (c7f7675). The existing test_top_k_eq_vocab_no_filtering already covers the k == V no-op case.

**Comment by JC-ut0** [comment]
/cherry-pick releases/v0.26.0rc
[Bot]: cherry-pick completed successfully. New PR created: https://github.com/vllm-project/vllm-ascend/pull/14537

File: benchmarks/ops/bench_topk_topp_sampler.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+            npu_vals = [r[2]["median"] if r[2] else 0 for r in results]
+
+            x = np.arange(len(names))
+            width = 0.35
+
+            ax.bar(x - width / 2, pt_vals, width, label="A: pytorch_npu", color="#4C72B0", alpha=0.85)
+            ax.bar(x + width / 2, npu_vals, width, label="B: npu_direct", color="#55A868", alpha=0.85)
+
+            ax.set_ylabel("Median time (ms)")
+            ax.set_title(title)
```
Comment (ningjingbengxiaohai) [inline benchmarks/ops/bench_topk_topp_sampler.py]: 删除这个文件

**Comment by JC-ut0** [comment]
/cherry-pick releases/v0.26.0rc
[Bot]: cherry-pick branch has been refreshed. Existing PR: https://github.com/vllm-project/vllm-ascend/pull/14537

**Review by ningjingbengxiaohai — APPROVED** [review]
